in-progress/7228-unified-bbjs-interface.md

```diff
@@ -0,0 +1,243 @@
+
+Native implementation (does not support `ClientIVC, AVM`)
+
+- `bb proof_as_fields | proof_as_fields_honk` (assumes `_honk` supports both Ultra and Mega)
+
+Wasm implementation (does not support `ClientIVC, AVM, MegaHonk`)
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirSerializeProofIntoFields | acirProofAsFieldsUltraHonk`
+
+#### `getVerificationKey(circuit: Circuit, flavor: Flavor): { vk: Buffer, vkAsFields: Fields }`
+
+Computes the verification key for a given circuit. Returns the verification key in binary and as fields, which allows the client to read its hash and the number of public inputs for the circuit. Users should call this once per circuit they want to verify, and reuse it for verifying as many proofs as needed.
+
+:question: Can we avoid returning `vkAsFields`, and convert from binary to fields in ts-land?
+
+Native implementation (`ClientIVC, AVM` not supported)
+
+- `bb write_vk | write_vk_ultra_honk | write_vk_mega_honk`
+- `bb vk_as_fields | vk_as_fields_ultra_honk | vk_as_fields_mega_honk`
+
+Wasm implementation (only `UltraPlonk` supported)
+
+- `acirGetCircuitSizes`
+- `commonInitSlabAllocator(size)`
+- `srsInitSrs(size)`
+- `acirInitProvingKey`
+- `acirGetVerificationKey`
+- `acirSerializeVerificationKeyIntoFields`
+
+#### `verify(proof: Buffer, vk: Buffer, flavor: Flavor): { isValid: boolean }`
+
+Verifies the given proof, which includes public inputs, against the given verification key. Both proof and vk are in raw binary format. Requires the output of `getVerificationKey` for the given circuit first.
+
+Native implementation (`ClientIVC` not supported)
+
+- `bb verify | avm_verify | verify_ultra_honk | verify_mega_honk`
+
+Wasm implementation (`ClientIVC, MegaHonk, AVM` not supported)
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirLoadVerificationKey`
+- `acirVerifyProof | acirVerifyUltraHonk`
+
+#### `getSolidityVerifier(vk: Buffer, flavor: Flavor): { contract: Buffer }`
+
+Generates a Solidity verifier contract for the given verification key in raw binary format. Only `UltraPlonk` is supported. Expected to be called once per circuit during development. Requires the output of `getVerificationKey`.
+
+Native implementation:
+
+- `bb contract`
+
+Wasm implementation:
+
+- `srsInitSrs(Crs(1))`
+- `acirNewAcirComposer(0)`
+- `acirLoadVerificationKey`
+- `acirGetSolidityVerifier`
+
+#### `getGateCount(circuit: Circuit, flavor: Flavor): { gates: number }`
+
+Counts the number of gates for a given circuit.
+
+Native implementation:
+
+- `bb gates`
+
+Wasm implementation:
+
+- `acirGetCircuitSizes`
+
+:question: `bb` exposes a single method to get gates with just a boolean "honk recursion". Does this mean that for `UltraPlonk` we can get circuit sizes setting this flag to false, and for `UltraHonk` and `MegaHonk` we set it to true? How about `ClientIVC`? And for `AVM` does it even make sense to ask about gate count?
```

Comment:
This is a hack I've introduced temporarily while Noir figures out how to correctly modify the verify_proof opcode so it can be used to verify proofs of different proof systems. It's currently set to true for UltraHonk only, so that programs that have verify_proof opcodes will link to the UltraHonk recursive verifier when we prove with UltraHonk.